In [1]:
import pandas as pd

data = {
    "question": [
        "What is the capital of France?",
        "What is the capital of England?",
        "What is the capital of Italy?",
        "What is the capital of Germany?",
        "What is the capital of Japan?",
        "What is the capital of Pakistan?",
        "What is the largest planet?",
        "What is the smallest planet?",
        "Which planet is known as the Red Planet?",
        "Which planet do we live on?",
        "What is the star closest to Earth?",
        "What is the largest ocean?",
        "What is the largest continent?",
        "What is the fastest land animal?",
        "What is the tallest animal?",
        "What animal is known as the king of the jungle?",
        "What color is grass?",
        "What color is the sky?",
        "What color is snow?",
        "How many days are in a week?",
        "How many months are in a year?",
        "What is two plus two?",
        "What is five plus five?",
        "What is ten minus five?",
        "What is three times three?",
        "Who invented the telephone?",
        "Who discovered penicillin?",
        "Who discovered gravity?",
        "What gas do humans breathe?",
        "What gas do plants absorb?",
        "What organ pumps blood?",
        "What organ helps us think?",
        "What is H2O?",
        "What is HCl?",
        "What is the hardest natural substance?",
        "What is the largest desert?",
        "What is the highest mountain?",
        "What is the smallest ocean?",
        "What is the currency of Japan?",
        "What is the currency of America?"
    ],

    "answer": [
        "Paris",
        "London",
        "Rome",
        "Berlin",
        "Tokyo",
        "Islamabad",
        "Jupiter",
        "Mercury",
        "Mars",
        "Earth",
        "Sun",
        "Pacific",
        "Asia",
        "Cheetah",
        "Giraffe",
        "Lion",
        "Green",
        "Blue",
        "White",
        "Seven",
        "Twelve",
        "Four",
        "Ten",
        "Five",
        "Nine",
        "Bell",
        "Fleming",
        "Newton",
        "Oxygen",
        "Carbon",
        "Heart",
        "Brain",
        "Water",
        "Acid",
        "Diamond",
        "Sahara",
        "Everest",
        "Arctic",
        "Yen",
        "Dollar"
    ]
}

df = pd.DataFrame(data)

In [2]:
df.head()

,question,answer
0,What is the capital of France?,Paris
1,What is the capital of England?,London
2,What is the capital of Italy?,Rome
3,What is the capital of Germany?,Berlin
4,What is the capital of Japan?,Tokyo


# Tokenize The Text

In [3]:
def tokenize(text):
    text=text.lower()
    text=text.replace("?","")
    text=text.replace("'","")
    return text.split() #Split into words (tokens)

In [4]:
tokenize("my name is khan what is yours?")

['my', 'name', 'is', 'khan', 'what', 'is', 'yours']

## When building the vocabulary, we reserve ID 0 for unknown words.
<UNK> stands for Unknown. so if unknown word found in input we want to predict from training it will be unk

In [5]:
vocab = {'<UNK>':0}

## build_vocab() takes a question-answer pair, tokenizes the text, and adds every new word to the vocabulary with a unique integer ID so the NLP model can later convert words into numbers.

In [6]:
def build_vocab(row):
    tokenised_question=tokenize(row['question'])
    tokenised_answer=tokenize(row['answer'])

    merged_tokens=tokenised_question + tokenised_answer

    for token in merged_tokens: #for token in merged_tokens:Loops through each word in that combined list.vocab[token] = len(vocab)
        #Assigns a unique integer index to the word based on the current size of vocab.
        # If vocab is empty (len = 0), the first word gets assigned index 0.The next word gets index 1, the next gets 2, and so on.
        if token not in vocab:
            vocab[token] = len(vocab)

In [7]:
vocab

{'<UNK>': 0}

apply() means:

"Take this function and run it on the DataFrame."

## It effectively does:

build_vocab(row1)

- how → 1
- are → 2
- you → 3
- i → 4
- am → 5
- fine → 6

In [8]:
df.apply(build_vocab, axis=1)#axis=1 means work row by row.Each row is passed into:buildvocab()

0     None
1     None
2     None
3     None
4     None
5     None
6     None
7     None
8     None
9     None
10    None
11    None
12    None
13    None
14    None
15    None
16    None
17    None
18    None
19    None
20    None
21    None
22    None
23    None
24    None
25    None
26    None
27    None
28    None
29    None
30    None
31    None
32    None
33    None
34    None
35    None
36    None
37    None
38    None
39    None
dtype: object

In [9]:
len(vocab)

119

In [10]:
print(vocab)

{'<UNK>': 0, 'what': 1, 'is': 2, 'the': 3, 'capital': 4, 'of': 5, 'france': 6, 'paris': 7, 'england': 8, 'london': 9, 'italy': 10, 'rome': 11, 'germany': 12, 'berlin': 13, 'japan': 14, 'tokyo': 15, 'pakistan': 16, 'islamabad': 17, 'largest': 18, 'planet': 19, 'jupiter': 20, 'smallest': 21, 'mercury': 22, 'which': 23, 'known': 24, 'as': 25, 'red': 26, 'mars': 27, 'do': 28, 'we': 29, 'live': 30, 'on': 31, 'earth': 32, 'star': 33, 'closest': 34, 'to': 35, 'sun': 36, 'ocean': 37, 'pacific': 38, 'continent': 39, 'asia': 40, 'fastest': 41, 'land': 42, 'animal': 43, 'cheetah': 44, 'tallest': 45, 'giraffe': 46, 'king': 47, 'jungle': 48, 'lion': 49, 'color': 50, 'grass': 51, 'green': 52, 'sky': 53, 'blue': 54, 'snow': 55, 'white': 56, 'how': 57, 'many': 58, 'days': 59, 'are': 60, 'in': 61, 'a': 62, 'week': 63, 'seven': 64, 'months': 65, 'year': 66, 'twelve': 67, 'two': 68, 'plus': 69, 'four': 70, 'five': 71, 'ten': 72, 'minus': 73, 'three': 74, 'times': 75, 'nine': 76, 'who': 77, 'invented': 78

### Vocabulary

* `<UNK>` $\rightarrow$ `0`
* `hello` $\rightarrow$ `1`
* `world` $\rightarrow$ `2`

---

### Input: `"hello aliens"`

1. **Split into words:** `["hello", "aliens"]`
2. **`"hello"`:** Found in vocab $\rightarrow$ add `1`
3. **`"aliens"`:** Not found $\rightarrow$ add `<UNK>` (`0`)

---

### Result: `[1, 0]`

In [11]:
def text_to_indices(text,vocab):
    indexed_text=[]
    for token in tokenize(text):
        if token in vocab:
            indexed_text.append(vocab[token])
        else:
            indexed_text.append(vocab['<UNK>'])
    return indexed_text

In [12]:
text_to_indices("my name is khan",vocab)

[0, 0, 2, 0]

## "How are you?"
       ↓
- tokenize()
       ↓
- ['how', 'are', 'you']
       ↓
- vocab lookup is it in vocab
       ↓
- [1, 2, 3] yes 

## output 
- [1, 2, 3]

sentence → words → vocabulary numbers

In [13]:
text_to_indices("How many days do we have in a week?",vocab)

[57, 58, 59, 28, 29, 0, 61, 62, 63]

In [14]:
import torch
from torch.utils.data import Dataset, DataLoader

In [ ]:
class qadataset(Dataset):
    def __init__(self,df,vocab):
        self.df=df       #df → your question-answer DataFrame
        self.vocab=vocab #vocab → your word-to-number dictionary

    def __len__(self):
        return self.df.shape[0]
    
    def __getitem__(self,index):
        numerical_question=text_to_indices(self.df.iloc[index]['question'],self.vocab) #Row Retrieval of both: Grabs the row at position index using .iloc[index].
        numerical_answer=text_to_indices(self.df.iloc[index]['answer'],self.vocab)

        return torch.tensor(numerical_question), torch.tensor(numerical_answer)


# So now we have 
# Question → [1, 2, 3]
# Answer   → [4, 5, 6]

- DataFrame
    ↓
- QADataset
    ↓
- "How are you?"
    ↓
- tokenize()
    ↓
- ['how', 'are', 'you']
    ↓
- text_to_indices()
    ↓
- [1, 2, 3]
    ↓
- torch.tensor()
    ↓
- tensor([1, 2, 3])
    ↓
- DataLoader
    ↓
- RNN / LSTM / Transformer

In [16]:
dataset=qadataset(df,vocab)

In [17]:
dataloader=DataLoader(dataset,batch_size=1,shuffle=True)

#### question=tensor([[ 555,  489, 2346, 2268,  490]]) answer tensor([1731, 1731, 2346, 2268, 1540,  555, 1958,  555])

In [18]:
for question,answer in dataloader:
    print(question,answer[0])

tensor([[ 1,  2,  3, 18, 37]]) tensor([38])
tensor([[1, 2, 3, 4, 5, 6]]) tensor([7])
tensor([[57, 58, 65, 60, 61, 62, 66]]) tensor([67])
tensor([[ 1,  2,  3, 18, 39]]) tensor([40])
tensor([[ 1,  2, 74, 75, 74]]) tensor([76])
tensor([[  1,   2,   3, 105, 106, 107]]) tensor([108])
tensor([[  1,   2, 103]]) tensor([104])
tensor([[57, 58, 59, 60, 61, 62, 63]]) tensor([64])
tensor([[23, 19,  2, 24, 25,  3, 26, 19]]) tensor([27])
tensor([[ 1, 50,  2,  3, 53]]) tensor([54])
tensor([[ 1,  2,  3, 21, 19]]) tensor([22])
tensor([[  1,   2,   3, 115,   5,  14]]) tensor([116])
tensor([[ 1,  2,  3,  4,  5, 10]]) tensor([11])
tensor([[ 1,  2, 68, 69, 68]]) tensor([70])
tensor([[77, 81, 82]]) tensor([83])
tensor([[ 1, 50,  2, 55]]) tensor([56])
tensor([[ 1,  2,  3,  4,  5, 14]]) tensor([15])
tensor([[ 1, 93, 94, 95]]) tensor([96])
tensor([[ 1,  2, 71, 69, 71]]) tensor([72])
tensor([[ 1, 50,  2, 51]]) tensor([52])
tensor([[23, 19, 28, 29, 30, 31]]) tensor([32])
tensor([[77, 81, 84]]) tensor([85])
tenso

In [19]:
import torch.nn as nn

In [ ]:
class simplernn(nn.Module):
    def __init__(self,vocab_size):
        super().__init__()
        self.embedding=nn.Embedding(vocab_size,embedding_dim=50)
        self.rnn=nn.RNN(50,64,batch_first=True)
        self.fc=nn.Linear(64,vocab_size)# fc stands for Fully Connected layer.Takes the 64-number summary of the sentence 
                                        # and projects it across the entire vocabulary so each word gets a score.

    def forward(self,question):
        embedded_question=self.embedding(question)#Converts words from plain IDs (like [4, 12, 9]) into lists of 50 numbers each
        hidden,final=self.rnn(embedded_question)
        output=self.fc(final.squeeze(0))

        return output

### Concrete Example Walkthrough

Let's say our vocabulary has only **5 words**:

| Word | ID |
| --- | --- |
| `<PAD>` | 0 |
| `how` | 1 |
| `are` | 2 |
| `you` | 3 |
| `good` | 4 |

So, `vocab_size = 5`.

#### Step 1: Input

You send in the question: `"how are you"`

As IDs, this is: `[1, 2, 3]`

#### Step 2: Embedding

* `1` (`how`) $\rightarrow$ list of 50 numbers
* `2` (`are`) $\rightarrow$ list of 50 numbers
* `3` (`you`) $\rightarrow$ list of 50 numbers

#### Step 3: RNN Reads

* Reads `how` $\rightarrow$ updates memory
* Reads `are` $\rightarrow$ updates memory
* Reads `you` $\rightarrow$ finishes reading.
* Result (`final`): **one summary list of 64 numbers**.

#### Step 4: Linear Layer (`fc`)

The layer takes those 64 numbers and outputs 5 scores (one for each word in the vocabulary):

```text
Score for <PAD>: -1.2
Score for how:    0.4
Score for are:   -0.1
Score for you:    0.2
Score for good:   4.8   <-- Highest score!

```

The model picks **`good`** (ID 4) as its predicted answer.

In [21]:
learning_rate = 0.001
epochs = 20

In [29]:
print(len(vocab))

119


# The model internally sets up:len(vocab)
- self.embedding = nn.Embedding(119, 50)  # Table of 1,000 words -> 50 numbers each
- self.rnn       = nn.RNN(50, 64, ...)     # Reads the 50 numbers, keeps 64 in memory
- self.fc        = nn.Linear(64, 1000)     # Outputs 1,000 scores (one for each word)

In [22]:
model = simplernn(len(vocab))

In [23]:
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)

c:\Users\Furqan Khan\AppData\Local\miniconda3\envs\agent_env2\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [24]:
# training loop

for epoch in range(epochs):

  total_loss = 0

  for question, answer in dataloader:

    optimizer.zero_grad()

    # forward pass
    output = model(question)

    # target: answer sequence
    # output shape: (batch, vocab_size)
    # answer[:, 0] is only the first word, so we use the whole answer sequence
    # for a simpler fix, model should output one logit per token class, but this notebook
    # is set up as a single-token classifier. Keep the existing structure and train on the first answer token.
    loss = criterion(output, answer[:, 0])

    # gradients
    loss.backward()

    # update
    optimizer.step()

    total_loss = total_loss + loss.item()

  print(f"Epoch: {epoch+1}, Loss: {total_loss:4f}")


Epoch: 1, Loss: 191.270746
Epoch: 2, Loss: 167.919292
Epoch: 3, Loss: 146.880384
Epoch: 4, Loss: 125.806308
Epoch: 5, Loss: 105.299480
Epoch: 6, Loss: 86.395931
Epoch: 7, Loss: 70.484784
Epoch: 8, Loss: 56.574122
Epoch: 9, Loss: 44.879207
Epoch: 10, Loss: 35.821037
Epoch: 11, Loss: 28.721276
Epoch: 12, Loss: 23.052439
Epoch: 13, Loss: 18.782736
Epoch: 14, Loss: 15.509155
Epoch: 15, Loss: 13.018520
Epoch: 16, Loss: 11.078172
Epoch: 17, Loss: 9.521440
Epoch: 18, Loss: 8.272638
Epoch: 19, Loss: 7.269127
Epoch: 20, Loss: 6.492599


In [25]:
def predict(model, question, threshold=0.5):

  # convert question to numbers
  numerical_question = text_to_indices(question, vocab)

  # tensor
  question_tensor = torch.tensor(numerical_question).unsqueeze(0)

  # send to model
  output = model(question_tensor)

  # convert logits to probs
  probs = torch.nn.functional.softmax(output, dim=1)

  # find index of max prob
  value, index = torch.max(probs, dim=1)

  if value < threshold:
    print("I don't know")

  print(list(vocab.keys())[index])

In [30]:
# Assuming you have a test_loader or train_loader:
# Use the existing DataLoader variable: dataloader
model.eval()
correct = 0
total = 0

with torch.no_grad():
    for questions, answers in dataloader:
        predictions = model(questions).argmax(dim=1)
        targets = answers[:, 0]

        correct += (predictions == targets).sum().item()
        total += targets.numel()

acc = 100 * correct / total
print(f"Model Accuracy: {acc:.2f}%")
print(f"Model Accuracy: {acc:.2f}%")

Model Accuracy: 100.00%
Model Accuracy: 100.00%


100 because over dataset is very small

In [26]:
predict(model, "capital of japan")

tokyo


In [27]:
df.head()

,question,answer
0,What is the capital of France?,Paris
1,What is the capital of England?,London
2,What is the capital of Italy?,Rome
3,What is the capital of Germany?,Berlin
4,What is the capital of Japan?,Tokyo


In [28]:
list(vocab.keys())[1]

'what'

### RNN-Based Question Answering System — Summary

#### Goal

Build a simple NLP model that can answer a question by predicting a single-word answer.

Example:

**Question:**

> What is the capital of France?

**Answer:**

> Paris

---

### Workflow

#### 1. Load Dataset

The dataset contains question-answer pairs.

| Question                       | Answer  |
| ------------------------------ | ------- |
| What is the capital of France? | Paris   |
| What is the largest planet?    | Jupiter |

---

#### 2. Tokenization

Convert text into words.

```text
What is AI?
```

↓

```python
["what", "is", "ai"]
```

---

#### 3. Build Vocabulary

Assign an ID to every unique word.

```python
{
    "<UNK>":0,
    "what":1,
    "is":2,
    "paris":3
}
```

---

#### 4. Convert Text to Numbers

Neural networks cannot understand words.

```text
what is france
```

↓

```python
[1, 2, 5]
```

---

#### 5. Create Dataset & DataLoader

Prepare question-answer pairs for training.

```python
(question_tensor, answer_tensor)
```

---

#### 6. Embedding Layer

Convert word IDs into dense vectors.

```python
1
```

↓

```python
[0.12, -0.45, 0.88, ...]
```

---

#### 7. RNN Layer

Reads words one by one and remembers context.

```text
what → is → the → capital → of → france
```

The final hidden state represents the meaning of the whole question.

---

#### 8. Linear Layer

Uses the RNN output to predict the answer word.

```text
Question
   ↓
Embedding
   ↓
RNN
   ↓
Linear
   ↓
Predicted Answer
```

---

#### 9. Training

For each question:

1. Forward pass
2. Calculate loss
3. Backpropagation
4. Update weights

The model gradually learns the correct answers.

---

#### 10. Prediction

When a new question is given:

```text
What is the largest planet in our solar system?
```

↓

```text
Jupiter
```

(if it learned that pattern from training data)

---

### Architecture

```text
Question
   ↓
Tokenization
   ↓
Vocabulary Mapping
   ↓
Embedding Layer
   ↓
RNN Layer
   ↓
Linear Layer
   ↓
Answer Word
```

---

### What You Learn from This Project

✅ Tokenization

✅ Vocabulary Building

✅ Text-to-Index Conversion

✅ Custom PyTorch Dataset

✅ Embeddings

✅ RNN Basics

✅ Training Loop

✅ Question Answering Pipeline

---

### Limitation

This model predicts only **one word** as the answer.

Example:

```text
Q: What is the capital of France?
A: Paris
```

It cannot generate long answers like:

```text
Paris is the capital city of France.
```

For that, you need:

* LSTM + Attention
* Seq2Seq
* Transformers
* BERT
* GPT

### One-Line Summary

**This project teaches how to convert text into numbers, process it with an RNN, and predict a one-word answer from a question using PyTorch.**
